src/transformStateWithClones.js

```diff
@@ -6,8 +6,51 @@
  *
  * @return {Object[]}
  */
-function transformStateWithClones(state, actions) {
-  // write code here
+function transformStateWithClones(state = {}, actions = []) {
+  const clonedState = JSON.parse(JSON.stringify(state));
+  const stateHistory = [];
+
+  for (const action of actions) {
+    switch (action.type) {
+      case 'addProperties':
+        addProperties(clonedState, action.extraData, stateHistory);
+        break;
+      case 'removeProperties':
+        removeProperties(clonedState, action.keysToRemove, stateHistory);
+        break;
+      case 'clear':
+        clearProperties(clonedState, stateHistory);
+        break;
+    }
+  }
+}
+
+function addProperties(clonedState, extraData, stateHistory) {
+  Object.assign(clonedState, extraData);
+
+  const addHistory = { ...clonedState };
+
+  stateHistory.push(addHistory);
+}
+
+function removeProperties(clonedState, keysToRemove, stateHistory) {
+  for (const key of keysToRemove) {
+    delete clonedState[key];
+  }
+
+  const deleteHictory = { ...clonedState };
+
+  stateHistory.push(deleteHictory);
+}
+
+function clearProperties(clonedState, stateHistory) {
+  for (const key in clonedState) {
+    delete clonedState[key];
+  }
+
+  const clearHictory = { ...clonedState };
+
+  stateHistory.push(clearHictory);
 }
```

Comment:
The function name 'clearProperties' does not accurately reflect that it also handles updating the state history. Consider renaming it to something more descriptive, such as 'applyClearPropertiesAndRecordHistory'.